src/zarr/core/array.py

```diff
@@ -4149,3 +4354,48 @@ def _parse_deprecated_compressor(
     elif zarr_format == 2 and compressor == compressors == "auto":
         compressors = ({"id": "blosc"},)
     return compressors
+
+
+def _parse_data_params(
+    *,
+    data: np.ndarray[Any, np.dtype[Any]] | None,
+    shape: ShapeLike | None,
+    dtype: npt.DTypeLike | None,
+) -> tuple[np.ndarray[Any, np.dtype[Any]] | None, ShapeLike, npt.DTypeLike]:
+    """
+    Ensure an array-like ``data`` parameter is consistent with the ``dtype`` and ``shape``
+    parameters.
+    """
+    if data is None:
+        if shape is None:
+            msg = (
+                "The data parameter was set to None, but shape was not specified. "
+                "Either provide a value for data, or specify shape."
+            )
+            raise ValueError(msg)
+        if dtype is None:
+            msg = (
+                "The data parameter was set to None, but dtype was not specified."
+                "Either provide an array-like value for data, or specify dtype."
+            )
+            raise ValueError(msg)
+
+    else:
+        if shape is not None:
+            msg = (
+                "The data parameter was used, but the shape parameter was also "
+                "used. The shape parameter will be ignored, and the data.shape "
+                "attribute will be used instead."
+            )
+            warnings.warn(msg, UserWarning, stacklevel=2)
```

Comment:
I would just error here instead

Reply:
added error behavior in https://github.com/zarr-developers/zarr-python/pull/2761/commits/7b30744ea184103d7590f5f0a98bb5f13dbb673d